### PR TITLE
Fix crash when parsing valid CycloneDX formats missing optional metadata fields

### PR DESCRIPTION
**Pull request by kurt-r2c**
# Description of the PR

Fixes https://github.com/guacsec/guac/issues/169.

As of 0d557cbe66b42053e72d5f25c1bba94c905dcede, `guac` still crashes on the example from #169 (referenced also in #199): https://github.com/JupiterOne/graph-github/blob/43fa5c43078eb16fad904fd1567133f9c296ea7f/reports/bom.json#L6-L11

image

The root cause is this line of code: https://github.com/guacsec/guac/blob/cb2b6fe5f6449d5225f1e3f55ad5f2d4741dc5d5/pkg/ingestor/parser/cyclonedx/parser_cyclonedx.go#L205

The CycloneDX specification as of 1.4 does not *require* the `component` field in `metadata`, nor does it *require* the BOM ref field: https://cyclonedx.org/docs/1.4/json/#metadata_component_bom-ref.

`guac` was encountering a NPE when trying to read these uninitialized fields from the CycloneDX struct.

Some SBOM generators that only operate on package lockfiles (such as https://pypi.org/project/cyclonedx-bom/) do not generate the `component` metadata field, nor are they required to by the spec.

This change makes the existence of this field optional and "fails closed" to the `nil` handler introduced by https://github.com/guacsec/guac/pull/603 - I believe this preserves the **intent** of the design while gracefully handling the non-existence of this metadata field.

The test I added is not great - it does check that a schema without these fields in metadata *parses* but the predicate output is `nil` due to the existing handlers, which I did not want to make sweeping changes to. I leave this up to the maintainers to decide how this *should* behave.

# PR Checklist

- [ ] All commits have [a Developer Certificate of Origin (DCO)](https://wiki.linuxfoundation.org/dco) -- they are generated using `-s` flag to `git commit`.
- [ ] All new changes are covered by tests
- [ ] If GraphQL schema is changed, `make generate` has been run
- [ ] If `collectsub` protobuf has been changed, `make proto` has been run
- [ ] All CI checks are passing (tests and formatting)
- [ ] All dependent PRs have already been merged

### CODE REVIEW

File: pkg/ingestor/parser/cyclonedx/parser_cyclonedx.go [hunk 1]
```diff
@@ -202,7 +202,13 @@ func (c *cyclonedxParser) GetPredicates(ctx context.Context) *assembler.IngestPr
 
 	preds := &assembler.IngestPredicates{}
 
-	toplevel := c.getPackageElement(string(c.cdxBom.Metadata.Component.BOMRef))
+	var toplevel []*model.PkgInputSpec = nil
+
+	if c.cdxBom.Metadata.Component != nil && c.cdxBom.Metadata.Component.BOMRef != "" {
```
Comment (pxp928) [inline pkg/ingestor/parser/cyclonedx/parser_cyclonedx.go:207]: Hello @kurt-r2c! Thanks for the PR! This is something that warrants further discussion on how it should be handled in GUAC.

Comment (lumjjb) [reply pkg/ingestor/parser/cyclonedx/parser_cyclonedx.go:207]: I am wondering what the use case of having a CDX document without the top level component is for... if it is the case where it is interpretted as here's a list of all the components I use, then it should parse to a bunch of singletons, however, if it is not , then my thought would be to either use the generic `guac/cdx/<name>` or to just reject it. let's make sure we have this on the priority for triage.

**Comment by pxp928** [comment]
Hey @kurt-r2c, could you elaborate a bit more on your use case here? Based on the PR, it would generate `component` nodes without a top level `component` field in `metadata`. This would result in a disconnected node without a parent. How would you tie this information back to the package that the SBOM is being generated for?

**Comment by kurt-r2c** [comment]
@pxp928 I was just trying to ingest a Python CycloneDX SBOM generated for a project using https://pypi.org/project/cyclonedx-bom/ to start exploring the tool. The SBOM output from `cyclonedx-bom` is valid according to the spec - if guac needs additional context to function, it needs to explicitly require some additional arguments/constraints.

**Comment by lumjjb** [comment]
Yea let's perhaps add a message and link to this issue for discussion in the warning message. In case someone else comes along with a use case and a way to handle it.

**Comment by pxp928** [comment]
Thanks @kurt-r2c for opening the PR and bringing this issue to our attention. We have created issue #976 to track this and also PR https://github.com/guacsec/guac/pull/992 to ensure that GUAC fails with an error message. Closing this PR as it it no longer needed.

**Pull request closed without merging**